virtcontainers/qemu.go

```diff
@@ -837,6 +853,64 @@ func (q *qemu) hotplugRemoveCPUs(amount uint32) (uint32, error) {
 	return amount, q.sandbox.storage.storeHypervisorState(q.sandbox.id, q.state)
 }
 
+func (q *qemu) hotplugMemory(slot uint32, memMB uint64, op operation) error {
+	// We do not support memory hot unplug.
+	if op == removeDevice {
+		return errors.New("cannot hot unplug memory device")
+	}
+
+	// rule out extremely large parameter
+	if memMB > math.MaxInt32 {
+		return fmt.Errorf("cannot hotplug %d MB memory: size too large", memMB)
+	}
+
+	maxMem, err := q.hostMemMB()
+	if err != nil {
+		return err
+	}
+
+	// calculate current memory
+	currentMemory := int(q.config.DefaultMemSz)
+	if q.vmConfig.Memory > 0 {
+		currentMemory = int(q.vmConfig.Memory)
+	}
+	currentMemory += q.state.HotpluggedMemory
+
+	// Don't exceed the maximum amount of memory
+	if currentMemory+int(memMB) > int(maxMem) {
+		return fmt.Errorf("Unable to hotplug %d MB memory, currently this SB has %d memory and the maximum amount is %d",
+			memMB, currentMemory, q.config.DefaultMemSz)
+	}
+
+	return q.hotplugAddMemory(slot, int(memMB))
+}
+
+func (q *qemu) hotplugAddMemory(slot uint32, amount int) error {
```

Comment:
the qmp function takes an ID I wonder if for clarity would be better name it `id` or to if want to keep it refer that indeed we are adding an slot.

Reply:
The idea is to convert the slot number to an ID, either we do it in this function, or we need to do it somewhere else.

Reply:
I am fine with any case. I ask because maybe  a reader can think that is it is slot number where the memory will be inserted in the VM. This maybe not true all the time, qemu  inserts it into the first empty slot.